### PR TITLE
sentry-top v2

### PR DESCRIPTION
**Pull request by mattrobenolt**
top

This brings a version of Sentry Top natively into Sentry which doesn't
rely on tsdb and only leverages ephemeral data. Since we're not relying
on tsdb to source our data, we can pull data that we weren't tracking
before, such as events sent to project ids that don't exist.

Sentry Top works by setting a key, sentry:top, which specifies the
IP/port for Sentry to send data to. When shut down, the key is deleted.
This allows the ability to remotely attach into a running session, then
detach without any overhead at runtime.

Only one `sentry top` instance may be running at one time.

This will be expanded to provide active filters, to say, "only show me
event.blacklisted" events, or, "only show me events rejected because bad
auth", etc.

/cc @getsentry/infrastructure

TODO:
- [ ] Add new signal into Sentry which checks for `sentry:top` key.
- [ ] Make sure `sentry:top` key access is always fast and doesn't require a round tip to network for every check.
- [ ] Sampling rate? (I could manage 20k+/s on my Macbook Air, but is _more_ data more useful than say, 50% and extrapolating?)

### CODE REVIEW

File: src/sentry/runner/commands/top.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+from collections import defaultdict, deque
+from threading import Thread
+from simplejson import loads
+
+import click
+from sentry.runner.decorators import configuration
+
+
+@click.command()
+@click.option('--bind', default='0.0.0.0:18000', help='Bind address.', metavar='ADDRESS')
```
Comment (mattrobenolt) [inline src/sentry/runner/commands/top.py]: Not sure of what port to use here. :/

File: src/sentry/runner/commands/top.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+        except ValueError:
+            broadcast_ip, broadcast_port = broadcast, None
+
+    if broadcast_port is None:
+        broadcast_port = bind_port
+
+    broadcast_port = int(broadcast_port)
+
+    if not force:
+        current = options.get('sentry:top')
```
Comment (dcramer) [inline src/sentry/runner/commands/top.py]: Weren't we removing options with the : ?

File: src/sentry/runner/commands/top.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    def __init__(self, size=1):
+        self._size = size
+        # deques should be 1 larger than expected since
+        # we need a window. The latest bucket is always being
+        # filled up, and the others are used for data.
+        self._times = deque([0], size + 1)
+        self._buckets = deque([0], size + 1)
+
+    def incr(self):
+        ts = int(time())
```
Comment (dcramer) [inline src/sentry/runner/commands/top.py]: I think you want a ring buffer here

Comment (mattrobenolt) [inline src/sentry/runner/commands/top.py]: Isn't that effectively what I have here with the `deque`? It's fixed size memory.

Comment (dcramer) [inline src/sentry/runner/commands/top.py]: it has some similar properties but its not the same

with a ring buffer you dont store arbitrary buckets, you only store the last N and they're ordered/fixed

**Comment by ye** [comment]
@mattrobenolt I like that GIF animation you posted up there. Care to share what tool you use to capture that? Sorry it's a little off topic. 

Btw, this top-alike feature is awesome!

**Comment by mattrobenolt** [comment]
@ye I use [LICEcap](http://www.cockos.com/licecap/).

**Comment by markstory** [comment]
This pull request cannot be merged currently. Without further updates it will be closed in a few days.

**Pull request closed without merging**